src/Features/CSharp/Portable/CodeRefactorings/UseExplicitType/UseExplicitTypeCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,148 @@
+
+            var declaration = GetDeclaration(root, textSpan);
+            Debug.Assert(declaration.IsKind(SyntaxKind.VariableDeclaration, SyntaxKind.ForEachStatement, SyntaxKind.DeclarationExpression));
+
+            TypeSyntax declaredType = s_useExplicitTypeHelper.CanOfferFix(declaration, semanticModel, cancellationToken);
+            if (declaredType == null)
+            {
+                return;
+            }
+
+            State state = State.Generate(declaration, semanticModel, optionSet,
+                isVariableDeclarationContext: declaration.IsKind(SyntaxKind.VariableDeclaration), cancellationToken: cancellationToken);
+
+            // UseExplicitType analyzer/fixer already gives an action in this case
+            if (s_useExplicitTypeHelper.IsStylePreferred(semanticModel, optionSet, state, cancellationToken))
+            {
+                return;
+            }
+
+            Debug.Assert(state != null, "analyzing a declaration and state is null.");
+            if (!s_useExplicitTypeHelper.TryAnalyzeVariableDeclaration(declaredType, semanticModel, optionSet, cancellationToken))
+            {
+                return;
+            }
+
+            MakeRefactoring(context, document, declaredType, textSpan, cancellationToken);
+        }
+
+        private static SyntaxNode GetDeclaration(SyntaxNode root, TextSpan textSpan)
+        {
+            // Ex:
+            // var (x, y) = ...
+            // (var x, ...) = ...
+            // foreach (var (x, y) in ...) ...
+            var declarationExpression = root.FindToken(textSpan.Start).GetAncestor<DeclarationExpressionSyntax>();
+            if (declarationExpression != null)
+            {
+                return declarationExpression;
+            }
+
+            // Ex:
+            // var x = ...;
+            // var x;
+            // for (var x ... ) ...
+            // using (var x = ...) ...
+            var localDeclaration = root.FindToken(textSpan.Start).GetAncestor<VariableDeclarationSyntax>();
+            if (localDeclaration != null)
+            {
+                return localDeclaration;
+            }
+
+            // Ex:
+            // foreach (var x in ...) ...
+            var foreachStatement = root.FindToken(textSpan.Start).GetAncestor<ForEachStatementSyntax>();
+            if (foreachStatement != null)
+            {
+                return foreachStatement;
+            }
+
+            return null;
+        }
+
+        private static void MakeRefactoring(CodeRefactoringContext context, Document document, TypeSyntax typeName, TextSpan textSpan, CancellationToken cancellationToken)
+        {
+            if (!typeName.IsParentKind(SyntaxKind.DeclarationExpression))
+            {
+            }
+
+            if (typeName.OverlapsHiddenPosition(cancellationToken))
+            {
+                return;
+            }
+
+            if (!typeName.Span.IntersectsWith(textSpan.Start))
+            {
+                return;
+            }
```

Comment:
ah, here it is.  interesting way to do things :)

Reply:
Moved those two checks up to caller. That's clearer and saved some work too.

---
In reply to: [172283945](https://github.com/dotnet/roslyn/pull/25208#discussion_r172283945) [](ancestors = 172283945)